File: CSharp/ExtensibleCommands/ExtensibleCommandsUnitTests/AbortableSleepCommand.cs
using System.Threading;

using ExtensibleCommands;

namespace ExtensibleCommandsUnitTest
{
    /// <summary> 
    /// Implements sleep command that can be aborted at any time.
    /// </summary>
    public class AbortableSleepCommand : SimpleCommand
    {
        /// <summary> Sleep time (in msec). When it expires, command terminates. </summary>
        public int SleepTimeMsec { get; }

        /// <summary> Internal event to be used to detect aborts </summary>
        private readonly ManualResetEventSlim _eventAborted = new ManualResetEventSlim(false);

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="sleepTimeMsec">Sleep time (msec)</param>
        public AbortableSleepCommand(int sleepTimeMsec) : this(sleepTimeMsec, "Abortable Sleep") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="sleepTimeMsec">Sleep time (msec)</param>
        /// <param name="name">Command name</param>
        public AbortableSleepCommand(int sleepTimeMsec, string name)
        {
            Name = name;
            SleepTimeMsec = sleepTimeMsec;
        }

        /// <summary> Aborts execution </summary>
        public override void Abort()
        {
            _aborted = true;
            _eventAborted.Set();
        }

        /// <summary> 
        /// The body of the command execution. 
        /// Sleep for a specified period of time unless interrupted by Abort() earlier.
        /// </summary>
        protected override void Execute()
        {
            _eventAborted.Reset();
            _eventAborted.Wait(SleepTimeMsec);

            ProcessAbortAndPauseEvents();
        }
    }
}

File: Java/ExtensibleCommands/src/test/java/org/extensiblecommands/AbortableSleepCommand.java
package org.extensiblecommands;

/**
 * Implements sleep command that can be aborted at any time.
 */
public class AbortableSleepCommand extends SimpleCommand {
    /**
     * Internal event to be used to detect aborts
     */
    private final ManualResetEvent eventAborted = new ManualResetEvent(false);

    /**
     * Sleep time (in msec). When it expires, command terminates.
     */
    private final int sleepTimeMsec;

    /**
     * Constructor
     * @param sleepTimeMsec     Sleep time (msec)
     */
    public AbortableSleepCommand(int sleepTimeMsec) {
        this(sleepTimeMsec, "Abortable Sleep");
    }

    /**
     * Constructor
     * @param sleepTimeMsec     Sleep time (msec)
     * @param name              Command name
     */
    public AbortableSleepCommand(int sleepTimeMsec, String name) {
        this.name = name;
        this.sleepTimeMsec = sleepTimeMsec;
    }

    /**
     * @return       Sleep time (in msec). When it expires, command terminates.
     */
    public int getSleepTimeMsec() {
        return sleepTimeMsec;
    }

    /**
     * Aborts execution
     */
    @Override
    public void abort() {
        aborted = true;
        eventAborted.set();
    }

    /**
     * Sleep for a specified period of time unless interrupted by Abort() earlier.
     */
    @Override
    protected void execute() throws InterruptedException {
        eventAborted.reset();
        eventAborted.waitOne(sleepTimeMsec);
        
        processAbortAndPauseEvents();
    }
}
